text/3730-nit-lint-level.md

```diff
@@ -0,0 +1,372 @@
+- Rustc shows `nit`s by default, rather than hide them and require a separate opt-in mechanism for Cargo to see them
+  - Problem: People directly using rustc or using a third-party build tool may be inundated with these
+  - Lint levels may change without breaking compatibility
+  - If stabilize with no `nit`s or only downgrading `warn`s to `nit`s, then there will be no difference in the amount of output
+  - We already have a mechanism, let's not invent a new one
+- `nits` lint group is created to give rustc-only or third-party build tools a built-in way to control these besides ignoring them like Cargo
+- Cargo could print a message that `nit`s are present to help raise awareness of how to show them but they will likely always be present, so this is noise to experienced users.
+
+# Prior art
+[prior-art]: #prior-art
+
+IntelliJ IDEA has the following [lint levels](https://www.jetbrains.com/help/idea/configuring-inspection-severities.html):
+- Error: Syntax errors
+- Warning: Code fragments that might produce bugs or require enhancement
+- Weak Warning: Code fragments that can be improved or optimized (redundant code, duplicated code fragments, and so on)
+- Server Problem
+- Grammar Error
+- Typo
+- Consideration: Code fragments that can be improved. This severity level is not marked on the error stripe and does not have a default highlighting style, but you can select one from the list of existing styles or configure your own.
+- No highlighting (but fix available)
+
+LSP has the following [diagnostic levels](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#diagnostic):
+- Error
+- Warning
+- Information
+- Hint
+
+with the following diagnostic tags:
+- Unnecessary
+- Deprecated
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- `CARGO_BUILD_NITS=nit` is modeled off of `CARGO_BUILD_WARNINGS=warn` which is just saying "treat X level violations as Y level"
+  - `CARGO_BUILD_WARNINGS=warn` is "weird" but shouldn't show up too often as its the default
+  - However, the frequency is swapped with `CARGO_BUILD_NITS=nit`
+  - We may also want the terms to be consistent across the different level-control fields
+  - Reminder: `CARGO_BUILD_WARNINGS` is unstable and we can still change it
+
+## Naming
+
+The name should clearly communicate that there is no authoriative weight
+encouraging them to resolve them but that it is left to them and their reviewer
+to weigh whether to resolve or ignore them.
+
+The verb form is used for:
+- attribute
+- long flag
+
+Lint names are expected to be read with their level for what should be done, e.g.
+- `deny(let_and_return)` -> "deny the use of let-and-return in this code"
+- `warn(let_and_return)` -> "warn on the use of let-and-return in this code"
+
+The noun form is used for:
+- Dynamic lint group
+
+Options
+- `#[nit]` / `--nit` / `-N` / `-Wnits`
+  - Unsure how to word this as a sentence
```

Comment:
@joshtriplett 

> I think nit is a fine name.

Of course someone was ok with my bike-shed avoidance name :).  I guess I should have been more out there.

When I first switched to a company that used `nit` in code review, it felt odd and the intent wasn't to clear.  Was this a nitpick?  If they recognize it as such, why is it posted?

Also, not quite sure how to put to translate `#[nit(let_and_return)]` into a sentence in the same way as the other levels.

Reply:
That's fair: unlike `error`, `warn`, and `allow`, it isn't a verb.

Reply:
What about `hint`? Which is a verb, and similar to python type *hints*, you get a suggestion, and it's usually opt in to adopt.

In this case the IDE would give you hints. 

```cargo LEVEL=hint```

Reply:
@woile I've forked this thread over on the item for "hint" so we can keep this one focused.

Reply:
Feel free to resolve this thread in favor of the more general bikeshed. The point of this comment was "don't rule out `nit` as a possible name".

Reply:
nitpick is a verb.

Nitpick might be better, since reader may not be familiar with the use of "nit" as a shorthand for nitpick